pkg/image/controller/signature/signature_import_controller.go

```diff
@@ -0,0 +1,223 @@
+	controller := &SignatureImportController{
+		queue:          workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		ctx:            ctx,
+		fetchTimeout:   fetchTimeout,
+		imageClient:    imageClient,
+		imageLister:    imageInformer.Lister(),
+		imageHasSynced: imageInformer.Informer().HasSynced,
+	}
+
+	imageInformer.Informer().AddEventHandlerWithResyncPeriod(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			image := obj.(*imageapi.Image)
+			glog.V(4).Infof("Adding image %s", image.Name)
+			controller.enqueueImage(obj)
+		},
+		UpdateFunc: func(old, cur interface{}) {
+			image := cur.(*imageapi.Image)
+			glog.V(4).Infof("Updating image %s", image.Name)
+			controller.enqueueImage(cur)
+		},
+	}, resyncInterval)
+
+	return controller
+}
+
+func (s *SignatureImportController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer s.queue.ShutDown()
+
+	if !cache.WaitForCacheSync(stopCh, s.imageHasSynced) {
+		return
+	}
+
+	glog.V(5).Infof("Starting workers")
+	for i := 0; i < workers; i++ {
+		go wait.Until(s.worker, time.Second, stopCh)
+	}
+	<-stopCh
+	glog.V(1).Infof("Shutting down")
+
+}
+
+func (s *SignatureImportController) worker() {
+	for {
+		if !s.work() {
+			return
+		}
+	}
+}
+
+// work returns true if the worker thread should continue
+func (s *SignatureImportController) work() bool {
+	key, quit := s.queue.Get()
+	if quit {
+		return false
+	}
+	defer s.queue.Done(key)
+
+	err := s.syncImageSignatures(key.(string))
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("error syncing image %s, it will be retried: %v", key.(string), err))
+		s.queue.AddRateLimited(key)
+		return true
+	}
+
+	s.queue.Forget(key)
+	return true
+}
+
+func (s *SignatureImportController) enqueueImage(obj interface{}) {
+	_, ok := obj.(*imageapi.Image)
+	if !ok {
+		return
+	}
+	key, err := controller.KeyFunc(obj)
+	if err != nil {
+		glog.Errorf("Couldn't get key for object %+v: %v", obj, err)
+		return
+	}
+	s.queue.Add(key)
+}
+
+func (s *SignatureImportController) syncImageSignatures(key string) error {
+	image, err := s.imageLister.Get(key)
+	if err != nil {
+		glog.V(4).Infof("Unable to retrieve image %v: %v", key, err)
+		return err
+	}
+
+	currentSignatures, err := s.fetchImageSignatures(image)
+	if err != nil {
+		glog.V(4).Infof("Failed to fetch image %s signatures: %v", image.Name, err)
+		return err
+	}
+
+	t, err := kapi.Scheme.DeepCopy(image)
+	if err != nil {
+		return err
+	}
+	newImage := t.(*imageapi.Image)
+
+	// filteredSignatures contains only signatures that are not managed by this
+	// controller (manually added).
+	filteredSignatures := []imageapi.ImageSignature{}
+	for _, s := range newImage.Signatures {
+		if !isManagedSignature(s) {
+			filteredSignatures = append(filteredSignatures, s)
+		}
+	}
+
+	// Only add the signatures that we received from the signature store.
```

Comment:
@mtrmac @deads2k FYI

Basically, we are not going to touch anything that is not managed by this controller. We going to refresh the signatures that are managed, but ignore those that already exists (and are possible verified). If we don't receive signature that we already have and we manage it, we going to nuke it regardless of its verification status. I guess this is OK (iow. a signature storage remove the signature because it is no longer valid/etc. so there is no reason for keeping it around).

Reply:
I agree that local verification status should not be a factor in this.

Removing a signature just because it is not available on _one_ registry seems incorrect: AFAICT if the same image (manifest digest) is served by repo A, which serves signatures SA and S_common, and repo B, which serves signatures SB and S_common, the imported image’s signatures will keep flipping between (SA, S_common) and (SB, S_common), depending on which was the last repo synced. That doesn’t seem right; the imported image should have all of (SA, SB, S_common).

It’s clear enough that when A stops serving SA, the imported image should drop SA. It’s not obvious when S_common should be dropped. One of A/B may be the authoritative repo for S_common (or not _truly_ authoritative but a semi-trusted mirror of the authoritative repo), and dropping the signature by the authoritative repo would be a good reason to drop it everywhere; or it may be a quite untrusted third-party copy (something on Docker Hub perhaps), which should not be able to cause the cluster to drop S_common. I can only think of changing the system to store the existence of S_common not per-manifest-digest, but per-stream, but that’s probably a big change. On balance, I’d prefer never removing signatures to removing shared signatures when they should be kept around; the redesign, or perhaps some other solution, can hopefully be done later.

(Even if a signature has been dropped by the authoritative source, and is no longer valid / applicable, there _is_ a reason to keep the signature around for forensic investigations (“why was this backdoored image allowed to run on the cluster?”). But that’s clearly a lower-priority use case, and it is not necessarily best implemented in the registry: perhaps a per-node audit log would make more sense.)

Reply:
@mtrmac so to simplify it, we should not remove signatures that we already have but keep them on the image, right?

Reply:
That’s close enough for me, yes.